OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs

```diff
@@ -0,0 +1,142 @@
+using System.Collections.Generic;
+using System.IO;
+using OpenDreamRuntime.Procs;
+using OpenDreamRuntime.Resources;
+using OpenDreamShared.Dream;
+using OpenDreamShared.Resources;
+
+namespace OpenDreamRuntime.Objects.MetaObjects {
+    sealed class DreamMetaObjectIcon : DreamMetaObjectDatum
+    {
+        [Dependency] private readonly DreamResourceManager _rscMan = default!;
+
+        public enum DreamIconMovingMode : byte
+        {
+            Both = 0,
+            Movement = 1,
+            NonMovement = 2,
+
+        }
+
+        public static Dictionary<DreamObject, DreamIconObject> ObjectToDreamIcon = new();
+
+        public struct DreamIconObject {
+            // Actual DMI data
+            public DMIParser.ParsedDMIDescription Description; // TODO Eventually this should probably be removed in favor of just directly storing the data for the subset of the DMI that we actually care about
+
+            // These vars correspond to the args in icon/new() and the resulting /icon obj, not the actual DMI data
+            public string Icon;
+            public string? State; // Specific icon_state. Null is all states.
+            public AtomDirection? Direction; // Specific dir. Null is all dirs.
+            public byte? Frame; //1-indexed. Specific frame. Null is all frames.
+            public DreamIconMovingMode Moving;
+
+            public DreamIconObject(DreamResource rsc, DreamValue? state, DreamValue? dir, DreamValue? frame, DreamValue? moving)
```

Comment:
No uses of this ever pass `null` values, these DreamValues don't need to be nullable.